src/flint/fmpz.jl

```diff
@@ -143,17 +143,23 @@ end
 ################################################################################
 
 # Similar to hash for BigInt found in julia/base
+@inline function _fmpz_is_small(a::ZZRingElemOrPtr)
+  return __fmpz_is_small(data(a))
+end
 
-@inline function _fmpz_is_small(a::ZZRingElem)
-  return __fmpz_is_small(a.d)
+# "views" a non-small ZZRingElem as a readonly BigInt
+# the caller must ensure z is actually non-small, and
+# call GC.@preserve appropriately
+function _as_bigint(a::ZZRingElemOrPtr)
+  unsafe_load(Ptr{BigInt}(data(a) << 2))
 end
 
-@inline function _fmpz_limbs(a::ZZRingElem)
-  return __fmpz_limbs(a.d)
+@inline function _fmpz_size(a::ZZRingElemOrPtr)
+  return _fmpz_is_small(a) ? 1 : GC.@preserve a _as_bigint(a).size
 end
 
 function hash_integer(a::ZZRingElem, h::UInt)
-  return _hash_integer(a.d, h)
+  return GC.@preserve a _hash_integer(data(a), h)
```

Comment:
What do you need the preserve here for? a is not allowed to be a ptr by signature

Reply:
The return value of `data(a)` is only useful if `a` is preserved. (Unless `_fmpz_is_small(a)` is `true, then it it does not need to be preserved.)

Reply:
Indeed: `data(a)` returns an `Int` which however is a pointer in disguise. Julia could therefore decide to garbage collect `a` right after `data(a)` returned, but before `_hash_integer` starts. But then we may get garbage or a crash. Of course that only matters if the code calling `hash_integer` resp. `hash` does not access the value `a` anymore afterwards, which seems unlikely -- but not impossible.